<h1 id="strainfacts">StrainFacts</h1>
<p>StrainFacts is a “strain deconvolution” tool, which infers both strain genotypes and their relative abundance across samples directly from metagenotype data.</p>
<p>A “metagenotype” for a particular species in a particular sample is an multidimensional array counting the number of reads (from shotgun metagenomic sequencing, usually) that contained a particular allele at polymorphic sites in that species’s genome. Metagenotypes from multiple (maybe <em>many</em>) samples stacked together form the main input data-type analyzed by StrainFacts.</p>
<h4 id="citation">Citation</h4>
<p>For detailed information, check out the paper:</p>
<blockquote>
<p>B.J. Smith, X. Li, Z.J. Shi, A. Abate, K.S. Pollard. Scalable microbial strain inference in metagenomic data using StrainFacts. <em>Frontiers in Bioinformatics</em> (2022) doi:<a href="https://doi.org/10.3389/fbinf.2022.867386">10.3389/fbinf.2022.867386</a></p>
</blockquote>
<h2 id="installation">Installation</h2>
<p>The simplest possible installation directly from GitHub</p>
<pre><code>pip install git+https://github.com/bsmith89/StrainFacts.git#egg=sfacts</code></pre>
<p>However, installing <a href="#with-conda">as a conda environment</a> (described below) is probably the more robust option.</p>
<p>Installing PyTorch/Pyro with GPU support may be more challenging. Consider using a Docker/Singularity container like <a href="https://hub.docker.com/r/bsmith89/sfacts_dev">this one</a>, which has many of the necessary prerequisites (but excluding StrainFacts itself).</p>
<h2 id="examples">Examples</h2>
<p>Short vignettes explaining how to use StrainFacts for data processing, visualization, and evaluation are provided as Jupyter Notebooks:</p>
<ul>
<li><a href="https://byronjsmith.com/StrainFacts/simulate_data.html"><code>examples/simulate_data.ipynb</code></a></li>
<li><a href="https://byronjsmith.com/StrainFacts/filter_data.html"><code>examples/filter_data.ipynb</code></a></li>
<li><a href="https://byronjsmith.com/StrainFacts/fit_metagenotype.html"><code>examples/fit_metagenotype.ipynb</code></a></li>
<li><a href="https://byronjsmith.com/StrainFacts/fit_metagenotype_advanced.html"><code>examples/fit_metagenotype_advanced.ipynb</code></a></li>
<li><a href="https://byronjsmith.com/StrainFacts/evaluate_simulation_fit.html"><code>examples/evaluate_simulation_fit.ipynb</code></a></li>
</ul>
<h2 id="usage">Usage</h2>
<p>(<strong>NOTE</strong>: All of the files described in those notebooks and the usage examples below can be built automatically with Make e.g. try running: <code>make examples/sim.filt.fit.world.nc</code>.)</p>
<h3 id="get-help">Get help</h3>
<p>Get CLI help:</p>
<pre><code>sfacts --help  # List subcommands
sfacts fit --help  # Subcommand specific CLI usage</code></pre>
<h3 id="fit-a-model-to-metagenotype-data">Fit a model to metagenotype data</h3>
<p>Fit metagenotype data with 15 strains and default everything:</p>
<pre><code>sfacts fit \
    --verbose \
    --num-strains 15 \
    --random-seed 0 \
    sim.filt.mgen.nc sim.filt.fit.world.nc</code></pre>
<p>For the manuscript, StrainFacts was run with the following hyperparameters set explicitly:</p>
<pre><code>sfacts fit -m ssdd3_with_error  \
    --verbose --device cpu \
    --precision 32 \
    --random-seed 0 \
    --strains-per-sample 0.3 \
    --num-positions 5000 \
    --hyperparameters gamma_hyper=1e-10 \
    --hyperparameters pi_hyper=0.3 \
    --hyperparameters rho_hyper=0.5 \
    --hyperparameters alpha_hyper_mean=10.0 \
    --hyperparameters alpha_hyper_scale=1e-06 \
    --anneal-hyperparameters gamma_hyper=1.0 \
    --anneal-hyperparameters rho_hyper=5.0 \
    --anneal-hyperparameters pi_hyper=1.0 \
    --anneal-steps 10000 --anneal-wait 2000 \
    --optimizer-learning-rate 0.05 \
    --min-optimizer-learning-rate 1e-06 \
    --max-iter 1_000_000 --lag1 50 --lag2 100 \
    examples/sim.filt.mgen.nc examples/sim.filt.fit.world.nc</code></pre>
<p>In addition, values for <code>--strains-per-sample</code>/<code>--num-strains</code>, <code>--num-positions</code>, <code>--precision</code>, <code>--random-seed</code>, <code>--device</code> and input/output files were as described in <a href="#citation">the paper</a>.</p>
<h3 id="data-formats">Data Formats</h3>
<p>Intermediate files generated by StrainFacts use a NetCDF-based file format, and harness the <a href="https://docs.xarray.dev/en/stable/">xarray library</a> for reading and writing<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. Such files include input metagenotypes (suggested file extension <code>*.mgen.nc</code>), as well as more complex collections of both latent and observed variables aligned over their strain/sample/position axes (<code>*.world.nc</code>). All of these files are designed to be loaded and worked with using the <code>sfacts</code> Python library<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>.</p>
<p>For users who prefer to interact with StrainFacts via the command-line interface (CLI) and plain-text files, scripts are provided as subcommands to convert metagenotype TSVs to this format (see <code>sfacts load --metagenotype</code>). as well as to export key parameter estimates after fitting (see <code>sfacts dump</code>).</p>
<p>For example, to export tab-delimited relative abundance and genotypes tables from <code>examples/sim.filt.fit.world.nc</code>:</p>
<pre><code>sfacts dump examples/sim.filt.fit.world.nc \
    --genotype examples/sim.filt.fit.geno.tsv \
    --community examples/sim.filt.fit.comm.tsv</code></pre>
<pre><code>head sim.filt.fit.refit.geno.tsv sim.filt.fit.refit.comm.tsv

# OUTPUT:
#
# ==&gt; sim.filt.fit.refit.geno.tsv &lt;==
# strain    position    genotypes
# 0 0   0.046489883
# 0 1   0.038462497
# 0 2   9.877303e-07
# 0 3   7.4534853e-07
# 0 4   0.9227885
# 0 5   0.99933404
# 0 6   0.9999933
# 0 7   0.9999994
# 0 8   2.3873392e-06
# 
# ==&gt; sim.filt.fit.refit.comm.tsv &lt;==
# sample    strain  communities
# 0 0   0.00062094553
# 0 1   0.00034649053
# 0 2   0.00017691542
# 0 3   0.00060282537
# 0 4   0.000553946
# 0 5   0.0006001372
# 0 6   0.00032402115
# 0 7   0.6081786
# 0 8   0.00042524032</code></pre>
<h2 id="see-also">See Also</h2>
<h3 id="alternative-deconvolution-tools">Alternative deconvolution tools:</h3>
<ul>
<li><a href="https://github.com/cssmillie/StrainFinder">Strain Finder</a></li>
<li><a href="https://bitbucket.org/luo-chengwei/constrains/src/master/">ConStrains</a></li>
<li><a href="http://www.cs.ucf.edu/~xiaoman/mixtureS/">MixtureS</a></li>
<li><a href="https://github.com/chrisquince/DESMAN">DESMAN</a></li>
<li><a href="https://github.com/wshuai294/PStrain">PStrain</a></li>
</ul>
<p>Strain Finder is the most similar to StrainFacts, and works well for reasonably sized data. However, it can be very slow as the number of strains and samples gets large.</p>
<h3 id="metagenotypers">Metagenotypers</h3>
<ul>
<li><a href="https://github.com/zjshi/gt-pro">GT-Pro</a></li>
<li><a href="https://github.com/czbiohub/iggtools">MIDAS / IGGtools</a></li>
<li><a href="https://github.com/biobakery/metaphlan">StrainPhlan</a></li>
</ul>
<p>GT-Pro is the preferred metagenotyper for StrainFacts. While other metagenotypers will also work (e.g. MIDAS or StrainPhlan assuming their outputs are formatted correctly), only bi-allelic and core genome SNP sites are currently supported.</p>
<h3 id="uhgg">UHGG</h3>
<p>GT-Pro is based on the <a href="http://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_genomes/human-gut/v1.0/">UHGG v1.0</a> species reference database. This is a subset of the recently updated <a href="https://www.ebi.ac.uk/metagenomics/genome-catalogues/human-gut-v2-0">MGnify / UHGG v2.0</a> database. The MGnify website provides a convenient way to browse this reference.</p>
<h2 id="how-to-hack-on-strainfacts">How to Hack on StrainFacts</h2>
<h3 id="editable-installation">Editable Installation</h3>
<p>First, clone the repository:</p>
<pre><code>git clone https://github.com/bsmith89/StrainFacts.git
cd StrainFacts</code></pre>
<p>If you want to modify the example notebooks, configure the clean/smudge filter:</p>
<p><code>make .git_init</code></p>
<p>Then install with either conda or pip:</p>
<h4 id="with-conda">With <code>conda</code></h4>
<p>Build the environment:</p>
<pre><code>conda env create -n sfacts-dev -f envs/sfacts-dev.yaml
# Also see `make .conda` to run this command automatically.
conda activate sfacts-dev</code></pre>
<p>This includes an embedded <code>pip install --editable</code> for StrainFacts from the cloned repo.</p>
<h4 id="with-pip">With <code>pip</code></h4>
<pre><code>pip install -r requirements.pip
pip install -e .</code></pre>
<h3 id="integration-testing">Integration Testing</h3>
<pre><code>make clean
make test</code></pre>
<p>This runs a complex workflow, including</p>
<ul>
<li>simulation</li>
<li>data conversion</li>
<li>initial fitting</li>
<li>genotype refitting</li>
<li>evaluation of estimates against the simulated ground-truth</li>
</ul>
<h3 id="prototyping-in-jupyter">Prototyping in Jupyter</h3>
<pre><code>make start_jupyter</code></pre>
<h2 id="todo">TODO</h2>
<ul class="task-list">
<li><input type="checkbox" disabled="" checked="" />
A script for converting GT-Pro output TSVs to StrainFacts input data</li>
<li><input type="checkbox" disabled="" checked="" />
A script for extracting estimates from StrainFacts files to TSVs</li>
<li><input type="checkbox" disabled="" checked="" />
Complete installation instructions, as well as a robust artifact for running StrainFacts on various platforms (e.g. a Dockerfile or Conda environment specification)</li>
<li><input type="checkbox" disabled="" checked="" />
Example data with a tutorial/vignette for running StrainFacts and interpreting the output.<ul class="task-list">
<li><input type="checkbox" disabled="" checked="" />
Simple example fitting a metagenotype</li>
<li><input type="checkbox" disabled="" checked="" />
Flesh out the example with an explanation of all of the steps.</li>
<li><input type="checkbox" disabled="" />
<del>Example data from SRA with a GT-Pro step, instead of simulated</del></li>
<li><input type="checkbox" disabled="" />
Update README/Makefile/Examples to match new CLI interfaces.</li>
</ul></li>
<li><input type="checkbox" disabled="" checked="" />
The parameters selected for use in this paper (which we have found to be applicable across a range of datasets) will be set as the default.<ul>
<li>Caveat: All of the hyperparameter annealing parameters should still be set by the user</li>
</ul></li>
<li><input type="checkbox" disabled="" checked="" />
We will document some useful approaches for hyperparameter tuning.</li>
<li><input type="checkbox" disabled="" checked="" />
Improve the CLI documentation.<ul class="task-list">
<li><input type="checkbox" disabled="" />
Explicit blurb about how to run the model as in the paper</li>
<li><input type="checkbox" disabled="" />
Model description strings.</li>
</ul></li>
<li><input type="checkbox" disabled="" />
Refactor for the best looking code<ul class="task-list">
<li><input type="checkbox" disabled="" checked="" />
De-pluralize core datatypes (community not communities)</li>
<li><input type="checkbox" disabled="" checked="" />
Remove unused CLI apps/workflows/plotting code</li>
<li><input type="checkbox" disabled="" checked="" />
Use Python <code>logging</code> module instead of shoe-string <code>info()</code> function.</li>
<li><input type="checkbox" disabled="" />
Drop stale code in <code>sfacts.data</code>, <code>sfacts.evaluation</code>, and <code>sfacts.plot</code></li>
<li><input type="checkbox" disabled="" />
Simplify the model structure decorator.</li>
<li><input type="checkbox" disabled="" />
Fix mismatch between ‘gamma’/‘pi’ and ‘genotype’/‘community’. (Probably just change variable names in the model, yeah?)</li>
<li><input type="checkbox" disabled="" />
Move choice of strain numbers for fit and init apps to a set of helper functions.</li>
<li><input type="checkbox" disabled="" />
Refactor CLI app architecture to be more intuitive.</li>
</ul></li>
<li><input type="checkbox" disabled="" />
Memoization of expensive calculations on wrapped Datasets and DataArrays.</li>
</ul>
<section class="footnotes" role="doc-endnotes">
<hr />
<ol>
<li id="fn1" role="doc-endnote"><p>This has benefits for both computational and programmer efficiency.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2" role="doc-endnote"><p>It should be fairly straight-forward to load these NetCDF formatted files in other programming environments, but this has not been tested.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>
